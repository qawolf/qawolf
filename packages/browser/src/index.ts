// public API
export { BrowserContext, register } from "./ContextManager";
export { Page } from "./page/PageManager";
// TODO move to jest-playwright
export { launch, LaunchResult } from "./launch";

// internal API
export * from "./actions";
export * from "./keyboard";
export { ContextManager } from "./ContextManager";
export { PageManager } from "./page/PageManager";
