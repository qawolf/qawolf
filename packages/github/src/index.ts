import Octokit from "@octokit/rest";

const octokit = new Octokit({
  auth: "d9e1b19607b2b07ed5c6f2e78fa5fb1bd3cabf26"
});

const main = async () => {
  const res = await octokit.checks.create({
    owner: "qawolf",
    repo: "internal",
    name: "QA Wolf",
    head_sha: "9c52a7a6f2090c58bcb23ae066ce31920789f71c"
  });

  console.log("RESPONSE", res);
};

main();
