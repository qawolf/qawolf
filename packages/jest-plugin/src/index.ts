// setup fail-fast since we do not want to continue when a step fails
// hopefully jest-circus implements this behavior, but was not the case when I tested jest 24.9.0
// https://github.com/facebook/jest/issues/2867#issuecomment-476647864
// https://github.com/facebook/jest/issues/6527#issuecomment-506278251

type Callback = (name: string) => void;

const j = jasmine as any;

// require since there are no types
const failFast = require("@qawolf/jasmine-fail-fast");
const jasmineEnv = j.getEnv();
jasmineEnv.addReporter(failFast.init());
