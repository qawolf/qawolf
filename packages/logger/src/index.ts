import { Logger } from "./Logger";

export const browserLogger = new Logger("browser");

export const logger = new Logger("server");
