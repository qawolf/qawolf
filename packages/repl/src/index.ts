import { Reporter } from "./Reporter";

// export for types to be happy
export { repl, ReplWithContext } from "./ReplWithContext";
export { CONTEXT } from "./Context";

import { repl } from "./ReplWithContext";
import { CONTEXT } from "./Context";

// export with module.exports for Jest to be happy
const mainExport = Reporter as any;
mainExport.CONTEXT = CONTEXT;
mainExport.repl = repl;
module.exports = mainExport;
