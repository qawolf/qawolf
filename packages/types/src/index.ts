export type Action = "click" | "type" | "scroll" | "select";

export type Callback<S = void, T = void> = (data?: S) => T;

export type ElementDescriptor = {
  alt?: string | null;
  ariaLabel?: string | null;
  classList?: string[] | null;
  dataValue?: string | null;
  href?: string | null;
  iconContent?: string[] | null;
  id?: string | null;
  inputType?: string | null;
  isContentEditable?: boolean | null;
  labels?: string[] | null;
  name?: string | null;
  parentText?: string[] | null;
  placeholder?: string | null;
  src?: string | null;
  tagName?: string | null;
  innerText?: string | null;
  title?: string | null;
  xpath?: string | null;
};

export interface Event {
  name: EventName;
  isTrusted: boolean;
  pageId?: number;
  target: ElementDescriptor;
  time: number;
}

export type EventName =
  | "click"
  | "input"
  | "keydown"
  | "keyup"
  | "paste"
  | "scroll";

export interface InputEvent extends Event {
  name: "input";
  value: string | null;
}

export interface KeyEvent extends Event {
  name: "keydown" | "keyup";
  // https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code
  value: string;
}

export type Locator = {
  action: Action;
  dataAttribute: string | null;
  target: ElementDescriptor;
  timeoutMs: number;
  value?: string | null;
};

export interface PasteEvent extends Event {
  name: "paste";
  value: string;
}

export interface ScrollEvent extends Event {
  name: "scroll";
  value: ScrollValue;
}

export type ScrollValue = {
  x: number;
  y: number;
};

export type Size = "desktop" | "tablet" | "mobile";

export type Step = {
  action: Action;
  index: number;
  pageId?: number;
  target: ElementDescriptor;
  value?: StepValue;
};

export type StepValue = string | ScrollValue | null | undefined;

export type Workflow = {
  name: string;
  size: Size;
  steps: Step[];
  url: string;
};
