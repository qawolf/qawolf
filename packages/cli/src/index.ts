#!/usr/bin/env node

import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { yellow } from "kleur";
import program from "commander";
import { snakeCase } from "lodash";
import { saveCiTemplate } from "./ci";
import { record } from "./record";
import { test } from "./test";
import { parseUrl } from "./utils";
const { version } = require("../package");

program.version(version);

program.usage("<command> [options]");

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "save events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  logger.verbose(`record url "${url.href}"`);

  const name = snakeCase(optionalName || url.hostname!);
  await record(url, name, cmd.events);
});

program
  .command("test [name]")
  .description("run a test")
  .action(async optionalName => {
    await test(optionalName);
  });

program
  .command("azure")
  .description("set up an Azure Pipeline")
  .action(async () => {
    await saveCiTemplate("azure");
  });

program
  .command("circleci")
  .description("set up CircleCI")
  .action(async () => {
    await saveCiTemplate("circleci");
  });

program
  .command("github")
  .description("set up a GitHub Action")
  .action(async () => {
    await saveCiTemplate("github");
  });

program
  .command("gitlab")
  .description("set up GitLab CI/CD")
  .action(async () => {
    await saveCiTemplate("gitlab");
  });

program.arguments("<command>").action(cmd => {
  console.log(yellow(`Invalid command "${cmd}"\n`));
  program.help();
});

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
