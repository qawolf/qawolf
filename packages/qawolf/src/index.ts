#!/usr/bin/env node
import updateNotifier from "update-notifier";

const pkg = require("../package.json");

const isCLI = !module.parent;
if (isCLI) {
  updateNotifier({ pkg }).notify();

  require("@qawolf/cli");
}

// export public API
export { BrowserContext, launch, Page } from "@qawolf/browser";

export { repl } from "@qawolf/repl";

export { sleep, waitFor, waitUntil } from "@qawolf/web";

// make repl a global
import { repl } from "@qawolf/repl";
(global as any).repl = repl;
