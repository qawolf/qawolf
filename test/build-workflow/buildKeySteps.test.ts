import { loadFixtures } from '../loadFixtures';
import { buildKeySteps } from '../../src/build-workflow';

describe('buildKeySteps', () => {
  it('builds expected steps for login', async () => {
    const events = (await loadFixtures('login')).events;
    const steps = buildKeySteps(events);
    expect(steps).toMatchSnapshot();
  });

  it('builds expected steps for modifier keys', async () => {
    const events = (await loadFixtures('textInputsModifierKeys')).events;
    const steps = buildKeySteps(events);
    expect(steps).toMatchSnapshot();
  });

  it('builds expected steps for selectall', async () => {
    const events = (await loadFixtures('textInputsSelectAll')).events;
    const steps = buildKeySteps(events);
    expect(steps).toMatchSnapshot();
  });

  // TODO add test that page changes split steps
});
