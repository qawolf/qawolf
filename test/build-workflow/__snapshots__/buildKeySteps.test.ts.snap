// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildKeySteps builds expected steps for login 1`] = `
Array [
  Object {
    "action": "type",
    "event": Object {
      "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"\\" qaw_labels=\\"Username\\" /></form></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "autocomplete": "off",
          "id": "username",
          "qaw_labels": "Username",
          "type": "text",
          "value": "",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583106436649,
      "value": "t",
    },
    "index": 0,
    "value": "tomsmith",
  },
  Object {
    "action": "press",
    "event": Object {
      "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"username\\" type=\\"text\\" value=\\"tomsmith\\" qaw_labels=\\"Username\\" /></form></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "autocomplete": "off",
          "id": "username",
          "qaw_labels": "Username",
          "type": "text",
          "value": "tomsmith",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583106437954,
      "value": "Tab",
    },
    "index": 1,
    "value": "Tab",
  },
  Object {
    "action": "type",
    "event": Object {
      "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"password\\" type=\\"password\\" value=\\"\\" qaw_labels=\\"Password\\" /></form></div>",
      "isTrusted": true,
      "name": "paste",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "autocomplete": "off",
          "id": "password",
          "qaw_labels": "Password",
          "type": "password",
          "value": "",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583106438791,
      "value": "SuperSecretPassword!",
    },
    "index": 2,
    "value": "SuperSecretPassword!",
  },
  Object {
    "action": "press",
    "event": Object {
      "htmlSelector": "<div id=\\"root\\" qaw_innertext=\\"To log in, use tomsmith as the username, and SuperSecretPassword! as the password. Username Password Log in\\"><form qaw_innertext=\\"Username Password Log in\\"><input autocomplete=\\"off\\" id=\\"password\\" type=\\"password\\" value=\\"SuperSecretPassword!\\" qaw_labels=\\"Password\\" /></form></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "autocomplete": "off",
          "id": "password",
          "qaw_labels": "Password",
          "type": "password",
          "value": "SuperSecretPassword!",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583106439265,
      "value": "Enter",
    },
    "index": 3,
    "value": "Enter",
  },
]
`;

exports[`buildKeySteps builds expected steps for modifier keys 1`] = `
Array [
  Object {
    "action": "type",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704460692,
      "value": "!",
    },
    "index": 0,
    "value": "!@#$%^",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704471119,
      "value": "ArrowLeft",
    },
    "index": 1,
    "value": "ArrowLeft",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704471567,
      "value": "ArrowLeft",
    },
    "index": 2,
    "value": "ArrowLeft",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704471994,
      "value": "ArrowLeft",
    },
    "index": 3,
    "value": "ArrowLeft",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704472330,
      "value": "ArrowLeft",
    },
    "index": 4,
    "value": "ArrowLeft",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704472697,
      "value": "ArrowLeft",
    },
    "index": 5,
    "value": "ArrowLeft",
  },
  Object {
    "action": "press",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704473147,
      "value": "ArrowLeft",
    },
    "index": 6,
    "value": "ArrowLeft",
  },
  Object {
    "action": "type",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input\\" placeholder=\\"Text input\\" type=\\"text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input",
          "placeholder": "Text input",
          "type": "text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583704478537,
      "value": "~",
    },
    "index": 7,
    "value": "~™£¢",
  },
]
`;

exports[`buildKeySteps builds expected steps for selectall 1`] = `
Array [
  Object {
    "action": "fill",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input-filled']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input-filled\\" placeholder=\\"Filled text input\\" type=\\"text\\" value=\\"initial text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input-filled",
          "placeholder": "Filled text input",
          "type": "text",
          "value": "initial text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583107777030,
      "value": "h",
    },
    "index": 0,
    "value": "hotkeys",
  },
  Object {
    "action": "fill",
    "event": Object {
      "cssSelector": "[data-qa='html-text-input-filled']",
      "htmlSelector": "<div style=\\"display: flex; flex-wrap: wrap;\\" qaw_innertext=\\"Native HTML Material UI Text input Filled text input Password input Number input Textarea\\"><div class=\\"container\\" qaw_innertext=\\"Native HTML\\"><input data-qa=\\"html-text-input-filled\\" placeholder=\\"Filled text input\\" type=\\"text\\" value=\\"initial text\\" /></div></div>",
      "isTrusted": true,
      "name": "keydown",
      "page": 0,
      "target": Object {
        "attrs": Object {
          "data-qa": "html-text-input-filled",
          "placeholder": "Filled text input",
          "type": "text",
          "value": "initial text",
        },
        "children": Array [],
        "name": "input",
        "type": "tag",
        "voidElement": true,
      },
      "time": 1583107783358,
      "value": "c",
    },
    "index": 1,
    "value": "clickselectall",
  },
]
`;
